CDTDatastore/CDTDatastoreManager.m

```diff
@@ -55,72 +59,110 @@ - (CDTDatastore *)datastoreNamed:(NSString *)name
 {
-    //    if (![TD_Database isValidDatabaseName:name]) {
-    //      Not a public method yet
-    //    }
-
-    CDTDatastore *datastore = nil;
-
-    NSString *errorReason = nil;
-    TD_Database *db = [self.manager databaseNamed:name];
-    if (db) {
-        datastore = [[CDTDatastore alloc] initWithManager:self database:db encryptionKeyProvider:provider];
+    @synchronized (self) {
+        CDTDatastore *datastore = _openDatastores[name];
+        if (datastore != nil) {
+            NSLog(@"returning already open CDTDatastore %@", name);
+            return datastore;
+        }
+        NSLog(@"opening new CDTDatastore %@", name);
 
-        if (!datastore) {
-            errorReason = NSLocalizedString(@"Wrong key?", nil);
+        
+        NSString *errorReason = nil;
+        TD_Database *db = [self.manager databaseNamed:name];
+        if (db) {
+            datastore = [[CDTDatastore alloc] initWithManager:self database:db encryptionKeyProvider:provider];
+            
+            if (!datastore) {
+                errorReason = NSLocalizedString(@"Wrong key?", nil);
+            }
+        } else {
+            errorReason = NSLocalizedString(@"Invalid name?", nil);
+        }
+        
+        if (!datastore && error) {
+            NSDictionary *userInfo = @{
+                                       NSLocalizedDescriptionKey : NSLocalizedString(@"Couldn't create database.", nil),
+                                       NSLocalizedFailureReasonErrorKey : errorReason,
+                                       NSLocalizedRecoverySuggestionErrorKey : errorReason
+                                       };
+            *error = [NSError errorWithDomain:CDTDatastoreErrorDomain code:400 userInfo:userInfo];
+        }
+        
+        if (datastore != nil) {
+            _openDatastores[name] = datastore;
         }
-    } else {
-        errorReason = NSLocalizedString(@"Invalid name?", nil);
+        return datastore;
+        
     }
+}
 
-    if (!datastore && error) {
-        NSDictionary *userInfo = @{
-            NSLocalizedDescriptionKey : NSLocalizedString(@"Couldn't create database.", nil),
-            NSLocalizedFailureReasonErrorKey : errorReason,
-            NSLocalizedRecoverySuggestionErrorKey : errorReason
-        };
-        *error = [NSError errorWithDomain:CDTDatastoreErrorDomain code:400 userInfo:userInfo];
+- (void)closeDatastoreNamed:(NSString *)name {
+    @synchronized (self) {
+        NSLog(@"closing CDTDatastore %@", name);
+        CDTDatastore *ds = _openDatastores[name];
+        if (ds == nil) {
+            // this may not be an issue if delete was already called and it was removed there
+            NSLog(@"warning can't find CDTDatastore to close %@", name);
+            return;
+        }
+        [[ds database] close];
+        [_openDatastores removeObjectForKey:name];
     }
+}
 
-    return datastore;
+- (void)dealloc {
+    NSLog(@"dealloc CDTDatastoreManager");
+    /*
```

Comment:
forget to delete this? I assume the dealloc is called on databases as they fall out of scope.